imap/imap.go

```diff
@@ -0,0 +1,628 @@
+		for _, uid := range rsp.SearchResults() {
+			msgCount++
+			seq.AddNum(uid)
+		}
+	}
+
+	// nothing to request?! why you even callin me, foolio?
+	if seq.Empty() {
+		return
+	}
+
+	fCmd, err := imap.Wait(client.UIDFetch(seq, "INTERNALDATE", "BODY[]", "UID", "RFC822.HEADER"))
+	if err != nil {
+		responses <- Response{Err: fmt.Errorf("unable to perform uid fetch: %s", err)}
+		return
+	}
+
+	var email Email
+	for _, msgData := range fCmd.Data {
+		msgFields := msgData.MessageInfo().Attrs
+
+		// make sure is a legit response before we attempt to parse it
+		// deal with unsolicited FETCH responses containing only flags
+		// I'm lookin' at YOU, Gmail!
+		// http://mailman13.u.washington.edu/pipermail/imap-protocol/2014-October/002355.html
+		// http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch
+		if _, ok := msgFields["RFC822.HEADER"]; !ok {
+			continue
+		}
+
+		email, err = NewEmail(msgFields)
+		if err != nil {
+			responses <- Response{Err: fmt.Errorf("unable to parse email: %s", err)}
+			return
+		}
+
+		responses <- Response{Email: email}
+
+		if !markAsRead {
+			err = removeSeen(client, imap.AsNumber(msgFields["UID"]))
+			if err != nil {
+				responses <- Response{Err: fmt.Errorf("unable to remove seen flag: %s", err)}
+				return
+			}
+		}
+
+		if delete {
+			err = deleteEmail(client, imap.AsNumber(msgFields["UID"]))
+			if err != nil {
+				responses <- Response{Err: fmt.Errorf("unable to delete email: %s", err)}
+				return
+			}
+		}
+	}
+	return
+}
+
+func deleteEmail(client *imap.Client, UID uint32) error {
+	return alterEmail(client, UID, "\\DELETED", true)
+}
+
+func removeSeen(client *imap.Client, UID uint32) error {
+	return alterEmail(client, UID, "\\SEEN", false)
+}
+
+func alterEmail(client *imap.Client, UID uint32, flag string, plus bool) error {
+	flg := "-FLAGS"
+	if plus {
+		flg = "+FLAGS"
+	}
+	fSeq := &imap.SeqSet{}
+	fSeq.AddNum(UID)
+	_, err := imap.Wait(client.UIDStore(fSeq, flg, flag))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hasEncoding(word string) bool {
+	return strings.Contains(word, "=?") && strings.Contains(word, "?=")
+}
+
+func isEncodedWord(word string) bool {
+	return strings.HasPrefix(word, "=?") && strings.HasSuffix(word, "?=") && strings.Count(word, "?") == 4
+}
+
+func parseSubject(subject string) string {
+	if !hasEncoding(subject) {
+		return subject
+	}
+
+	dec := mime.WordDecoder{}
+	sub, _ := dec.DecodeHeader(subject)
+	return sub
+}
+
+// NewEmail will parse an imap.FieldMap into an Email. This
+// will expect the message to container the internaldate and the body with
+// all headers included.
+func NewEmail(msgFields imap.FieldMap) (Email, error) {
+	var email Email
+	// parse the header
+	var message bytes.Buffer
+	message.Write(imap.AsBytes(msgFields["RFC822.HEADER"]))
+	message.Write([]byte("\n\n"))
+	rawBody := imap.AsBytes(msgFields["BODY[]"])
+	message.Write(rawBody)
+	msg, err := mail.ReadMessage(&message)
```

Comment:
I'd recommend that we defer all the parsing below this line to the `github.com/jordan-wright/email` package we already use in Gophish. Specifically, the call to `email.NewEmailFromReader` appears to work, which returns an [`email.Email`](https://godoc.org/github.com/jordan-wright/email#Email) instance which has all the fields you'd expect (From, Subject, HTML, Body, etc.). This will allow us to remove a ton of parsing code in here, as well as multiple external dependencies.

Reply:
The one thing your email struct doesn't have is message UID. I [added this](https://github.com/jprobinson/eazye/blob/master/eazye.go#L188) to the `eazye` email package in order to delete campaign emails (or mark emails as unread that failed to report). Maybe you could add that field to your [email.Email](https://godoc.org/github.com/jordan-wright/email#Email) ?

Also if we're going down this route I wonder if it's worth rewriting the whole of `imap.go` with [go-imap](https://github.com/emersion/go-imap) and your `github.com/jordan-wright/email` 🤔

Reply:
I do agree that rewriting might be the best option. But to me the biggest hold up for this particular PR is the number of dependencies it adds just to handle the mail parsing. If we could remove the parsing functionality alone and instead use the existing `github.com/jordan-wright/email` then that'd be better, IMO. We can work on longer term rewrites later.

I mean, honestly if we think the effort is worth it and you'd be open to it, I don't see why we couldn't eventually make a new entire repo at `gophish/imap` that does just that - making IMAP more accessible. I could give write access so you could contribute to that as you'd see fit. Just an option to consider 🤷‍♂ 

For adding the UID, I can see the benefit. The reason I'd err away from having that in the email package (for now, until I've had more time to consider it) is that it's really meant to construct emails that are sent, whereas UID is an IMAP-specific field.

Perhaps the better short-term option would be to make a struct like this:

```
type IMAPEmail struct {
    UID          uint32          `json:"uid"`
    *email.Email
}
```

Then read the `*email.Email` from `NewEmailFromReader`, and use that to construct your `IMAPEmail`, which is used throughout this package for parsing. It'll still have all the other `email.Email` fields and functions, but with the addition of a `UID` field.

Reply:
Hi @jordan-wright, thanks for your patience on this - it's been hard to find dedicated time to properly sit down and tackle this. I've now implemented your email library and removed a lot of code that was imported from `eazye`. I've done basic testing and it works well; I haven't tested all error conditions yet but wanted to get it pushed out ASAP.

See: https://github.com/glennzw/gophish/commit/c6a629079fe1198c875a654d6b0e998f840b88a6